File: Mex/MDOF_FRF/MDOF_Eig_Visc_cpp/mexFunction.cpp
#include <complex>
#include <valarray>
#include "valarray_ComplexRealOperators.hpp"
#include "MexOperations.hpp"
#include "a_matrices.hpp"
#include "mex.hpp"
#include "mexAdapter.hpp"

using namespace std;
using namespace a_matrices;

using namespace matlab::data;
using matlab::mex::ArgumentList;

void MDOF_Eig_Visc(const Matrix<double>& M_mat,const Matrix<double>& C_mat,const Matrix<double>& K_mat, const bool& isPropotional, Matrix<complex<double> >& EigVectors_Normalized, valarray<complex<double> >& EigValues_vec);

class MexFunction : public matlab::mex::Function
{
	ArrayFactory factory;
	const std::shared_ptr<matlab::engine::MATLABEngine> matlabPtr = getEngine();

public:
	void operator()(ArgumentList outputs, ArgumentList inputs)
	{
		size_t N;
		checkArguments(outputs, inputs, N);

		Matrix<double> M_mat=ColMajor2RowMajor(static_cast<TypedArray<double>>(inputs[0]));
		Matrix<double> C_mat=ColMajor2RowMajor(static_cast<TypedArray<double>>(inputs[1]));
		Matrix<double> K_mat=ColMajor2RowMajor(static_cast<TypedArray<double>>(inputs[2]));

		bool isPropotional=false;
		if (inputs.size() == 4)
			isPropotional= inputs[3][0];

		//Do Actual Computations
		Matrix<complex<double> > EigVectors_Normalized(N,2*N);
		valarray<complex<double> > EigVal_vec(2*N);
		try
		{
			MDOF_Eig_Visc(M_mat,C_mat,K_mat,isPropotional, EigVectors_Normalized, EigVal_vec);
		}
		catch (char str[])
		{
			matlabPtr->feval(u"error", 0,
				std::vector<Array>({ factory.createScalar(str) }));
		}

		//Outputs
		outputs[0] = std::move(RowMajor2ColMajor(EigVectors_Normalized));
		if (outputs.size() > 1)
			outputs[1] = factory.createArray<complex<double> >({1, 2 * N }, &EigVal_vec[0], &EigVal_vec[2*N]);

		return;
	}

	void checkArguments(ArgumentList outputs, ArgumentList inputs, size_t &N)
	{
		if ((inputs.size() != 3) && (inputs.size() != 4))
			matlabPtr->feval(u"error", 0,
				std::vector<Array>({ factory.createScalar("Dear student, This function takes 3 or 4 inputs.") }));

		if ((outputs.size() != 1) && (outputs.size() != 2))
			matlabPtr->feval(u"error", 0,
				std::vector<Array>({ factory.createScalar("Dear student, this function returns one or two outputs only.") }));

		N = inputs[0].getDimensions()[0];
		if (N != inputs[0].getDimensions()[1])
			matlabPtr->feval(u"error", 0,
				std::vector<Array>({ factory.createScalar("Dear student, The M matrix must be square.") }));

		for (size_t n = 1; n < 3; n++)
		{
			if ((inputs[n].getDimensions()[0] != N) || (inputs[n].getDimensions()[1] != N))
				matlabPtr->feval(u"error", 0,
					std::vector<Array>({ factory.createScalar("Dear student, the C and K matrices must be square with size identical to M.") }));    //Improve this
		}

		if (inputs.size() == 4)
			if (inputs[3].getType() != ArrayType::LOGICAL || inputs[3].getNumberOfElements() != 1)
				matlabPtr->feval(u"error", 0,
					std::vector<Array>({ factory.createScalar("Dear student, isPropotional must be a logical scalar/integer.") }));
	};
};

File: Mex/MDOF_FRF/MDOF_Eig_Visc_Fortran/MEXFUNCTION.f90
#include <fintrf.h>

logical function mxGetLogicalScalar(mx)
    USE MX_INTERFACES

    implicit none
    mwPointer, intent(in) :: mx

    integer*1 integer1value(1)

    mxGetLogicalScalar = .false.
    IF ((mxIsLogical(mx) /= 0) .AND. (mxGetNumberOfElements(mx) == 1)) THEN
        CALL mxCopyPtrToInteger1(mxGetData(mx),integer1value,1)
        mxGetLogicalScalar= (integer1value(1) /= 0)
    END IF
end function

SUBROUTINE MEXFUNCTION(NLHS, PLHS, NRHS, PRHS)
! This subroutine is the main gateway to MATLAB.  When a MEX function
!  is executed MATLAB calls this subroutine

USE, INTRINSIC :: ISO_C_BINDING, ONLY: C_SIZE_T
USE IFPORT
USE MX_INTERFACES
USE MEX_INTERFACES

IMPLICIT NONE
#if defined(_WIN32) || defined(_WIN64)
!DEC$ ATTRIBUTES DLLEXPORT :: MEXFUNCTION
#else
!GCC$ ATTRIBUTES DLLEXPORT :: MEXFUNCTION
#endif
    MWPOINTER PLHS(*), PRHS(*)
    INTEGER(4) NLHS, NRHS
    !mwSize NN
    mwPointer N

    REAL(8), ALLOCATABLE :: M_mat(:,:),C_mat(:,:),K_mat(:,:)
    LOGICAL isProportional
    INTEGER(C_SIZE_T) NN
    COMPLEX(8), ALLOCATABLE :: EigVectors_Normalized(:,:), EigValues_vec(:)

    INTERFACE
        SUBROUTINE MDOF_Eig_Visc(M_mat,C_mat,K_mat,N,isProportional,EigValues_vec,EigVectors_Normalized)
            USE, INTRINSIC :: ISO_C_BINDING, ONLY: C_SIZE_T
            INTEGER(C_SIZE_T), INTENT(IN) :: N
            LOGICAL, INTENT(IN) :: isProportional
            REAL(8), INTENT(IN) ::  M_mat(N,N),C_mat(N,N),K_mat(N,N)
            COMPLEX(8), INTENT(OUT) :: EigVectors_Normalized(N,2*N), EigValues_vec(2*N)
	    END SUBROUTINE MDOF_Eig_Visc

        logical function mxGetLogicalScalar(mx)
            mwPointer, intent(in) :: mx
        end function mxGetLogicalScalar
    END INTERFACE

#if defined(_WIN32) || defined(_WIN64)
! For Windows only!
! This resets the floating point exception to allow divide by zero,
! overflow and invalid numbers. 
	INTEGER(2) CONTROL
	CALL GETCONTROLFPQQ(CONTROL)
	CONTROL = CONTROL .OR. FPCW$ZERODIVIDE
      CONTROL = CONTROL .OR. FPCW$INVALID
      CONTROL = CONTROL .OR. FPCW$OVERFLOW
	CALL SETCONTROLFPQQ(CONTROL)
#endif

    ! Check for proper number of arguments
    IF ((NRHS/=3) .AND. (NRHS/=4)) THEN
	    CALL mexErrMsgTxt('Dear student, This function takes 3 or 4 inputs.')
    END IF

    IF ((NLHS/=1) .AND. (NLHS/=2)) THEN
	    CALL mexErrMsgTxt('Dear student, this function returns one or two outputs only.')
    END IF

    N=mxGetM(PRHS(1))
    IF (N /= mxGetN(PRHS(1))) THEN
	    CALL mexErrMsgTxt('Dear student, The M_mat matrix must be square.')
    END IF

    DO NN=2,3
        IF ((mxGetM(PRHS(NN)) /= N) .OR. (mxGetN(PRHS(NN)) /= N)) THEN
	        CALL mexErrMsgTxt('Dear student, the C_mat and K_mat matrices must be square with size identical to M_mat.')    !Improve this
        END IF
    END DO

    IF (NRHS==4) THEN
        IF ((mxGetNumberOfElements(PRHS(4)) /= 1) .OR. (.NOT. mxIsLogical(PRHS(4)))) THEN
            CALL mexErrMsgTxt('Dear student, isProportional must be a logical scalar/integer.')
        END IF
    END IF

    !Inputs
    ALLOCATE(M_mat(N,N),C_mat(N,N),K_mat(N,N))
    CALL mxCopyPtrToReal8(mxGetDoubles(PRHS(1)), M_mat, N*N)
    CALL mxCopyPtrToReal8(mxGetDoubles(PRHS(2)), C_mat, N*N)
    CALL mxCopyPtrToReal8(mxGetDoubles(PRHS(3)), K_mat, N*N)
    isProportional=.FALSE.
    IF (NRHS==4) THEN
        isProportional=mxGetLogicalScalar(PRHS(4))
    END IF


    !Do Actual Computations
    ALLOCATE(EigVectors_Normalized(N,2*N), EigValues_vec(2*N))
    CALL MDOF_Eig_Visc(M_mat,C_mat,K_mat,N,isProportional,EigValues_vec,EigVectors_Normalized)

    !Outputs
    PLHS(1)=mxCreateDoubleMatrix(N,2*N,1)  !Create the output matrix
    CALL mxCopyComplex16ToPtr(EigVectors_Normalized, mxGetComplexDoubles(PLHS(1)), N*2*N)

    IF (NLHS > 1) THEN
        PLHS(2)=mxCreateDoubleMatrix(2*N,1,1)  ! Create the output matrix
    CALL mxCopyComplex16ToPtr(EigValues_vec, mxGetComplexDoubles(PLHS(2)), 2*N)
END IF

END SUBROUTINE MEXFUNCTION